### DISCUSSION TITLE
FastAPI return internal server error when response model validation fails

### DISCUSSION THREAD

**Discussion by JoeyG1973** [fastapi/fastapi#6122] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Body, Depends, FastAPI, HTTPException, Path, Query, Request
from pydantic import BaseModel, Field, ValidationError
from pprint import pprint as pp

app = FastAPI()


class MetadataOut(BaseModel):
    """
    Model for Metadata
    """

    region: str = Field(
        ...,
        title="region",
        description="region - one of dc00|dc01|dc02",
        regex=r"^(dc00|dc01|dc02)$",
    )


@app.get("/api/v1/metadata/fail", response_model=MetadataOut)
def metadatafail():
    return {"region": "whoops"}


@app.get("/api/v1/metadata/failwitherror", response_model=MetadataOut)
def metadatafailwitherror():
    try:
        return MetadataOut(region="whoops")
    except ValidationError as e:
        pp(e)
        raise HTTPException(status_code=422, detail=e.errors())


@app.get("/api/v1/metadata/pass", response_model=MetadataOut)
def metadatapass():
    return {"region": "dc01"}


if __name__ == "__main__":
    uvicorn.run(
        app,
        host="0.0.0.0",
        port=8080,
    )
```

### Description

I am asking this as a question because I am hoping that I just need to adjust my code a bit.

I am expecting when the response validation fails fastapi would return a 422 error and the pydantic descriptive error just like input works.

The only way I have gotten a half way decent return is to do what you see in the /api/v1/metadata/failwitherror endpoint.

### Operating System

Linux

### FastAPI Version

0.74.1

### Python Version

Python 3.8.10

**Comment by iudeen** [comment]
Can you provide full error trace? ie, Traceback

**Comment by JoeyG1973** [comment]
@iudeen 

```
INFO:     127.0.0.1:38430 - "GET /api/v1/metadata/fail HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/fastapi/applications.py", line 259, in __call__
    await super().__call__(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/routing.py", line 656, in __call__
    await route.handle(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/routing.py", line 259, in handle
    await self.app(scope, receive, send)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/starlette/routing.py", line 61, in app
    response = await func(request)
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/fastapi/routing.py", line 235, in app
    response_data = await serialize_response(
  File "/home/garciaj1/src/compute/metadata-api/.venv/lib/python3.8/site-packages/fastapi/routing.py", line 138, in serialize_response
    raise ValidationError(errors, field.type_)
pydantic.error_wrappers.ValidationError: 1 validation error for MetadataOut
response -> region
  string does not match regex "^(dc00|dc01|dc02)$" (type=value_error.str.regex; pattern=^(dc00|dc01|dc02)$)

```

**Comment by JarroVGIT** [comment, marked as answer]
This behaviour is intentionally and explained here in the docs:

https://fastapi.tiangolo.com/tutorial/handling-errors/#requestvalidationerror-vs-validationerror

TLDR: if your response cannot be casted as the pydantic model you've declared, that is considered a logic flaw and hence a 500 is thrown.

**Comment by JoeyG1973** [comment]
ok cool thank you

**Comment by JarroVGIT** [comment]
To be complete: a 422 is basically saying that the entity as provided by the client is unprocessable; so it's the clients fault. Rendering a response and incurring an error while doing that, is by definition a server error, and thus it will produce a HTTP 500. 

The way you've caught and handled the exception is an option (like in `metadatafailwitherror`) but it does violate the HTTP response codes (or actually, misuses them). That is up to you of course! 

If your response model is dependant on an input, you might want to cast it your self in a try/except block, rather than having FastAPI doing it by means of `response_model=`, and then return the object itself. 

If this answers your question, please close the issue.

**Comment by JoeyG1973** [comment]
@JarroVGIT unfortunately without the response_model= the swagger interface doesn't self-document the return.

**Comment by JoeyG1973** [comment]
@JarroVGIT thank you for answering so effectively.  :)

**Comment by JoeyG1973** [comment]
@JarroVGIT my turn to be complete :) this is the input:

```
@app.get("/api/v1/metadata", response_model=MetadataOut)
def metadata_without_client_ip_defined(request: Request):
    client_ip = IPv4Address(request.client.host)
    return get_metadata(client_ip)


@app.get("/api/v1/metadata/{client_ip}", response_model=MetadataOut)
def metadata_with_client_ip_defined(client_ip: IPv4Address):
    return get_metadata(client_ip)`

**Comment by tiangolo** [comment]
Thanks for the help here @iudeen and @JarroVGIT ! 👏  🙇 

Thanks for reporting back and closing the issue @JoeyG1973  👍

**Comment by torsami77** [comment]
another option is to move the said exception logic above/outside your try and exception block
